crates/matrix-sdk/src/send_queue/upload.rs

```diff
@@ -349,6 +621,134 @@ impl QueueStorage {
         Ok(())
     }
 
+    /// Consumes a finished gallery upload and queues sending of the final
+    /// gallery event.
+    #[cfg(feature = "unstable-msc4274")]
+    #[allow(clippy::too_many_arguments)]
+    pub(super) async fn handle_dependent_finish_gallery_upload(
+        &self,
+        client: &Client,
+        event_txn: OwnedTransactionId,
+        parent_key: SentRequestKey,
+        mut local_echo: RoomMessageEventContent,
+        item_infos: Vec<FinishGalleryItemInfo>,
+        new_updates: &mut Vec<RoomSendQueueUpdate>,
+    ) -> Result<(), RoomSendQueueError> {
+        // All uploads are ready: enqueue the event with its final data.
+        let sent_gallery = parent_key
+            .into_media()
+            .ok_or(RoomSendQueueError::StorageError(RoomSendQueueStorageError::InvalidParentKey))?;
+
+        let mut sent_media_vec = sent_gallery.accumulated;
+        sent_media_vec.push(AccumulatedSentMediaInfo {
+            file: sent_gallery.file,
+            thumbnail: sent_gallery.thumbnail,
+        });
+
+        let mut sent_infos = HashMap::<String, AccumulatedSentMediaInfo>::new();
+
+        for (item_info, sent_media) in zip(item_infos, sent_media_vec) {
+            let FinishGalleryItemInfo { file_upload: file_upload_txn, thumbnail_info } = item_info;
+
+            // Update cache keys in the cache store.
+            {
+                // Do it for the file itself.
+                let from_req = Media::make_local_file_media_request(&file_upload_txn);
+
+                // Store the sent media under the original cache key for later insertion into
+                // the local echo
+                sent_infos.insert(from_req.source.unique_key(), sent_media.clone());
+
+                trace!(from = ?from_req.source, to = ?sent_media.file, "renaming media file key in cache store");
+                let cache_store = client
+                    .event_cache_store()
+                    .lock()
+                    .await
+                    .map_err(RoomSendQueueStorageError::LockError)?;
+
+                // The media can now be removed during cleanups.
+                cache_store
+                    .set_ignore_media_retention_policy(&from_req, IgnoreMediaRetentionPolicy::No)
+                    .await
+                    .map_err(RoomSendQueueStorageError::EventCacheStoreError)?;
+
+                cache_store
+                    .replace_media_key(
+                        &from_req,
+                        &MediaRequestParameters {
+                            source: sent_media.file.clone(),
+                            format: MediaFormat::File,
+                        },
+                    )
+                    .await
+                    .map_err(RoomSendQueueStorageError::EventCacheStoreError)?;
```

Comment:
Any chance some of this can also be factored out, to avoid duplicating the code of `handle_dependent_finish_upload`?